flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/StreamingFileWriter.java

```diff
@@ -0,0 +1,180 @@
+
+	public StreamingFileWriter(
+			long bucketCheckInterval,
+			StreamingFileSink.BucketsBuilder<RowData, ?, ? extends
+					StreamingFileSink.BucketsBuilder<RowData, ?, ?>> bucketsBuilder,
+			FileSystemBucketListener listener) {
+		this.bucketCheckInterval = bucketCheckInterval;
+		this.bucketsBuilder = bucketsBuilder;
+		this.listener = listener;
+	}
+
+	@Override
+	public void initializeState(StateInitializationContext context) throws Exception {
+		super.initializeState(context);
+		final int subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
+		this.buckets = bucketsBuilder.createBuckets(subtaskIndex);
+
+		final OperatorStateStore stateStore = context.getOperatorStateStore();
+		bucketStates = stateStore.getListState(BUCKET_STATE_DESC);
+		maxPartCountersState = stateStore.getUnionListState(MAX_PART_COUNTER_STATE_DESC);
+
+		if (context.isRestored()) {
+			buckets.initializeState(bucketStates, maxPartCountersState);
+		}
+		inactivePartitions = new HashSet<>();
+		listener.setInactiveConsumer(b -> inactivePartitions.add(b));
+	}
+
+	@Override
+	public void snapshotState(StateSnapshotContext context) throws Exception {
+		super.snapshotState(context);
+		Preconditions.checkState(bucketStates != null && maxPartCountersState != null, "sink has not been initialized");
+		buckets.snapshotState(
+				context.getCheckpointId(),
+				bucketStates,
+				maxPartCountersState);
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		this.processingTimeService = getRuntimeContext().getProcessingTimeService();
+		long currentProcessingTime = processingTimeService.getCurrentProcessingTime();
+		processingTimeService.registerTimer(currentProcessingTime + bucketCheckInterval, this);
+	}
+
+	@Override
+	public void onProcessingTime(long timestamp) throws Exception {
+		final long currentTime = processingTimeService.getCurrentProcessingTime();
+		buckets.onProcessingTime(currentTime);
+		processingTimeService.registerTimer(currentTime + bucketCheckInterval, this);
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		super.processWatermark(mark);
+		this.currentWatermark = mark.getTimestamp();
+	}
+
+	@Override
+	public void processElement(StreamRecord<RowData> element) throws Exception {
+		buckets.onElement(
+				element.getValue(),
+				getProcessingTimeService().getCurrentProcessingTime(),
+				element.getTimestamp(),
+				currentWatermark);
+	}
+
+	@Override
+	public void notifyCheckpointComplete(long checkpointId) throws Exception {
+		super.notifyCheckpointComplete(checkpointId);
+		buckets.commitUpToCheckpoint(checkpointId);
+		CommitMessage message = new CommitMessage(
+				checkpointId,
+				getRuntimeContext().getIndexOfThisSubtask(),
+				getRuntimeContext().getNumberOfParallelSubtasks(),
+				new ArrayList<>(inactivePartitions));
+		output.collect(new StreamRecord<>(message));
+		inactivePartitions.clear();
+	}
+
+	@Override
+	public void close() throws Exception {
```

Comment:
Is it possible that there's some pending data between `close` and the last `notifyCheckpointComplete`?

Reply:
I'll modify to `endInput` and support commit pending data.